Add VSIM backend (Modelsim, Questasim)

johnsbrew opened the pull request:
Enable the use of Modelsim/Questasim as backend with chisel-tester

I am not really proud of adding suport for another commercial simulator but it is an important milestone for us to push forward the integration of chisel within our existing code base.
We indeed rely on lots of existing systemverilog libraries that are not supported with verilator.

I hope it will help other people as well.

I have started to test this new backend on my side but I have no idea how to integrate proper tests due to the very nature of vsim as closed-source commercial simulator.
Maybe use the same syntax as for vcs ?
`assume(firrtl.FileUtils.isVCSAvailable)`
I cannot say that it doesn't break vcs backend either as I cannot test it but other tests are still passing.

I think some of the changes of the vpi code could apply to other backends as well.

johnsbrew commented:
ping @chick

chick commented on src/main/resources/sim_api.h:
```diff
@@ -125,10 +127,17 @@ template <class T> class sim_api_t {
     pid_t pid = getpid();
+    char cwd[PATH_MAX];
+    if (getcwd(cwd, sizeof(cwd)) != NULL) {
+       printf("Current working dir: %s\n", cwd);
```
Adding a printf here makes me nervous, do you need it?

johnsbrew replied:
not really needed indeed, i will remove this line

chick commented on src/main/resources/sim_api.h:
```diff
@@ -125,10 +127,17 @@ template <class T> class sim_api_t {
-    in_ch_name  << std::dec << std::setw(8) << std::setfill('0') << pid << ".in";
-    out_ch_name << std::dec << std::setw(8) << std::setfill('0') << pid << ".out";
-    cmd_ch_name << std::dec << std::setw(8) << std::setfill('0') << pid << ".cmd";
+    in_ch_name  << cwd << "/" << std::dec << std::setw(8) << std::setfill('0') << pid << ".in";
```
Can you please say why you add the *cwd* to the names here.
I didn't notice but is something else changed to adjust for these new names

johnsbrew replied:
I will add the following comment in code:
`// As VSIM relative path management is a mess, enforce the use of absolute paths for all channels`

claassistantio commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/freechipsproject/chisel-testers?pullRequest=276) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/freechipsproject/chisel-testers?pullRequest=276) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/freechipsproject/chisel-testers?pullRequest=276) it.</sub>

johnsbrew commented:
hey @chick 
I addressed your concerns.
I also made many other small modifications as we started to use it for real and discovered some small issues.
